### PR TITLE
test-all-scream: Upgrade handling of baselines

### PR DESCRIPTION
**Pull request by jgfouca**
The default baseline commit should not be HEAD.
    
Additional minor cleanups.

### CODE REVIEW

File: components/scream/scripts/test_all_scream.py [hunk 1]
```diff
@@ -142,7 +172,7 @@ def generate_ctest_config(self, cmake_config, extra_configs, test):
         return result
 
     ###############################################################################
-    def generate_baselines(self, test, cleanup):
+    def generate_baselines(self, test):
```
Comment (jgfouca) [inline components/scream/scripts/test_all_scream.py]: @bartgol , why was the cleanup flag needed?

Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: Uhm, I'm looking at the history of test_all_scream.py. I think I wanted to avoid some files accidentally not being rebuilt when git switched the branch back to the current branch (from baseline).

I'm not sure if it's needed. In particular, I'm not 100% sure that git, when switching from branch A to branch B, changes the timestamp of `blah.cpp` if it is identical in the two branches. However, even if it doesn't, if one of the upstream deps has changed, the build of `blah.cpp.o` should be re-triggered anyways.

tl;dr: I think I was being overly pessimistic, and there is no need to wipe the build dir when switching branches.

Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: Wait. What if someone uses things like globbing in some cmake file. Are we sure rebuilds will always be correctly triggered? My guess is 'yes', since we re-run cmake every time.

File: components/scream/scripts/test-all-scream [hunk 2]
```diff
@@ -41,38 +41,33 @@ OR
...
+    parser.add_argument("--baseline-dir", default=None,
+        help="Use baselines from the given directory, skip baseline creation.")
+
+    parser.add_argument("-m", "--machine",
+        help="Provide machine name. This is required for internal kokkos builds and dashboard submission")
 
     parser.add_argument("--no-tests", action="store_true", help="Only build baselines, skip testing phase")
 
-    parser.add_argument("--keep-tree", action="store_true", help="Allow to keep the current work tree when testing against HEAD (only valid with `-b HEAD`)")
+    parser.add_argument("-d", "--keep-tree", action="store_true",
```
Comment (bartgol) [inline components/scream/scripts/test-all-scream]: I find it confusing when the single-letter flag is not the initial of the verbose flag. In this case, you cannot use `-k`, since that's already the short for `--kokkos`. I would do away with single-letter flag here. Besides, `-d` sounds like "delete" (although I'm assuming you meant 'dirty').

File: components/scream/scripts/test_all_scream.py (hunk 1 above)
Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: It looks like the cleanup was related to data, not source files.

Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: Whoops, everything except data.

Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: OK, I think the idea here is that, if the baseline commit is the same as the HEAD commit, there's no reason to rebuild scream when running the tests.

File: components/scream/scripts/test_all_scream.py [hunk 3]
```diff
@@ -73,12 +98,19 @@ def __init__(self, cxx, kokkos, submit, parallel, fast_fail, baseline_ref, basel
                 self._proc_count = 1
 
         if self._keep_tree:
-            if self._baseline_dir=="NONE":
+            expect(not is_repo_clean(), "Makes no sense to use --keep-tree when repo is clean")
```
Comment (bartgol) [inline components/scream/scripts/test_all_scream.py]: Sure, it makes no sense, but should we really error out? Perhaps just print a warning?

File: components/scream/scripts/test_all_scream.py [hunk 4]
```diff
@@ -45,9 +47,32 @@ def __init__(self, cxx, kokkos, submit, parallel, fast_fail, baseline_ref, basel
...
+        # Compute baseline info
+        import pdb
+        pdb.set_trace()
+        expect(not (self._baseline_ref and self._baseline_dir),
+               "Makes no sense to specify a baseline generation commit if using pre-existing baselines ")
+        if self._baseline_dir is None:
+            if self._baseline_ref is None:
+                # Compute baseline ref
+                if self._keep_tree:
+                    self._baseline_ref = "HEAD"
```
Comment (bartgol) [inline components/scream/scripts/test_all_scream.py]: It makes no sense to use the `--keep-tree` flag together with a baseline ref different from HEAD. In this series of if's, we process the `--keep-tree` flag only if there is no baseline selected. Could we check that first? I want to avoid some errors due to copy+paste, where `--keep-tree` is ignored because a baseline ref was specified, and the user is confused.

File: components/scream/scripts/test_all_scream.py [hunk 5]
```diff
@@ -45,9 +47,32 @@ def __init__(self, cxx, kokkos, submit, parallel, fast_fail, baseline_ref, basel
...
+                else:
+                    self._baseline_ref = get_common_ancestor("origin/master")
```
Comment (bartgol) [inline components/scream/scripts/test_all_scream.py]: This could be problematic. Say a user does not specify the baseline ref, and his/her branched forked from origin/master. The tests may pass, but that does not mean it's ok to merge. Even though the AT will test against master, and even though we should always specify the baseline ref, I find it safer to default to current origin/master, rather than the common ancestor. If you want to test against the common ancestor, you can figure out the SHA manually, then pass it to `-b`.

File: components/scream/scripts/test_all_scream.py [hunk 6]
```diff
@@ -170,29 +202,32 @@ def generate_baselines(self, test, cleanup):
...
     ###############################################################################
-    def generate_all_baselines(self, git_baseline_head, git_head):
+    def generate_all_baselines(self):
     ###############################################################################
-        print("Generating baselines for ref {}".format(git_baseline_head))
+        git_head_commit     = get_current_commit()
+        git_head_ref        = get_current_head()
+        git_baseline_commit = get_current_commit(commit=self._baseline_ref)
+
+        need_co = git_baseline_commit != git_head_commit
```
Comment (bartgol) [inline components/scream/scripts/test_all_scream.py]: I know you have `co` as an alias of `checkout` on your machine, but perhaps we should be verbose here, so whomever sees this file next understands what is actually needed.

File: components/scream/scripts/test_all_scream.py (hunk 1 above)
Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: I think you are correct; manually re-running cmake and make should be adequate to get a correct re-build of scream without needing to wipe the binary dir.

File: components/scream/scripts/test_all_scream.py (hunk 3 above)
Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: I think we should error out; the user is likely confused and is not testing what they think they are testing.

Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: Fair enough.

File: components/scream/scripts/test_all_scream.py (hunk 4 above)
Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: There's a check later in this function that, if keep_tree is set, it's OK for the repo to be dirty. Keep-tree doesn't really need to do anything else.

File: components/scream/scripts/test_all_scream.py (hunk 5 above)
Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: If the user forks from commit A, produces commit B, which is BFB with A, but in the meantime, commit C has been added to origin/master which is not BFB, why would we want to fail the test?

File: components/scream/scripts/test_all_scream.py (hunk 3 above)
Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: I'm not a python guru, so this may be naive: the expect is not indented inside the if. Is that a problem?

Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: The expect is part of the block that gets executed if self._keep_tree is true.

Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: Oh, sorry. I didn't realized the line above is red, hence removed.

**Comment by jgfouca** [comment]
@bartgol , I haven't tested this thoroughly yet, but see if you agree with the approach in the latest commit. Developers default to common-ancestor checking, integration testing (detected by presence of JENKINS_HOME env var) gets true integration testing.

**Comment by bartgol** [comment]
Nice. I can take care of updating the jenkins jobs once this PR goes in.

**Comment by jgfouca** [comment]
They should work as-is, right?

**Comment by jgfouca** [comment]
As long as the jenkins jobs are letting the tool select the baseline_ref, it should work without modifications (I think).

**Comment by bartgol** [comment]
> They should work as-is, right?

You're right. But I'd still like to make the jenkins jobs more self-explanatory, and pass the `-i` flag to test-all-scream.
